vcd/resource_vcd_network_pool.go

```diff
@@ -0,0 +1,524 @@
+package vcd
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"strings"
+)
+
+const (
+	networkProviderVcenter       = "vCenter"
+	networkProviderNsxtManager   = "NSX-T manager"
+	backingTypeTransportZone     = "Transport Zone"
+	backingTypePortGroup         = "Port Group"
+	backingTypeDistributedSwitch = "Distributed Switch"
+)
+
+var (
+	// supportedNetworkPoolTypes defines the network pool that we can create and modify
+	supportedNetworkPoolTypes = []string{
+		types.NetworkPoolGeneveType,    // GENEVE
+		types.NetworkPoolVlanType,      // VLAN
+		types.NetworkPoolPortGroupType, // PORTGROUP_BACKED
+	}
+	// networkProviders defines the appropriate network provider for each type of network Pool
+	networkProviders = map[string]string{
+		types.NetworkPoolGeneveType:    networkProviderNsxtManager,
+		types.NetworkPoolVlanType:      networkProviderVcenter,
+		types.NetworkPoolPortGroupType: networkProviderVcenter,
+		types.NetworkPoolVxlanType:     networkProviderVcenter, // NSX-V backed. Read-only
+	}
+)
+
+func resourceNetworkPoolBacking(origin string) *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Optional:    origin == "resource",
+				Computed:    true,
+				Description: "Backing name",
+			},
+			"id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Backing ID",
+			},
+			"type": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Backing Type (one of 'Transport Zone', 'Port Group', 'Distributed Switch')",
+			},
+		},
+	}
+}
+
+var resourceNetworkPoolVlanIdRange = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"start_id": {
+			Type:        schema.TypeInt,
+			Required:    true,
+			Description: "Start of the IDs range",
+		},
+		"end_id": {
+			Type:        schema.TypeInt,
+			Required:    true,
+			Description: "End of the IDs range",
+		},
+	},
+}
+
+func resourceVcdNetworkPool() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceNetworkPoolCreate,
+		ReadContext:   resourceNetworkPoolRead,
+		UpdateContext: resourceNetworkPoolUpdate,
+		DeleteContext: resourceNetworkPoolDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceNetworkPoolImport,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of network pool.",
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "Type of the network pool (one of `GENEVE`, `VLAN`, `PORTGROUP_BACKED`)",
```

Comment:
Just to spark a discussion, would `TRANSPORT_ZONE` fit better than `GENEVE` in this case? 
UI uses this:
```
Geneve backed 
Create a network pool backed by an NSX-T Transport Zone
```

But I'd say `TRANSPORT_ZONE` is more intuitive and we use it later?

cc @Didainius @adambarreiro @lvirbalas

Reply:
I use `GENEVE` because it's the most prominent word in the UI wizard. If I put `TRANSPORT_ZONE`, users may be taken aback and wonder what is the correspondence with the UI.

Reply:
I agree that `GENEVE` is what we see in the UI. Besides, `TRANSPORT_ZONE` is an ambiguous keyword - it could mean a lot of different things. `GENEVE` is the protocol that NSX-T uses.